Parser followup 2

LenAnderson opened the pull request:
**Overview for users: [V2-2.md](https://github.com/user-attachments/files/16404608/V2-2.md)**

---

- keep secondary auto complete values on screen when fully typed or auto completed
- add scope as second argument to enumProvider function (allows access to scoped vars)
- add auto complete support for multiple unnamed arguments (set `splitUnnamedArgument: true`)
- add `matchProvider` to auto complete options (custom matching function to be used instead of begins/includes/fuzzy)
- add `valueProvider` to auto complete options (custom value / replacer function when auto complete option is selected)
- add scoped variables to `commonEnumProviders.variables()`
- add variable hiding to `commonEnumProviders.variables()` (scoped hides chat hides global)
- add "any number" and "any variable" auto complete options to `/add` enumProvider
- trim whitespace from start of first part and from end of last part of unsplit unnamed args that are split due to content (e.g., includes immediate closure and text) to make it behave closer to regular unsplit args
- add crude debugger in modal QR editor
- export `evalBoolean` and `parseBooleanOperands`
- add `/break` to exit closure early and potentially break out of loops and similar commands, optional return value to override the last pipe value
- add `/break` support to `/times`
- add `/break` support to `/while`
- add `/break` support to `/run` / `/:`, acts like a return statement
- add command source indicator in auto complete details (core=favicon, core extension=favicon+green cubes, third party extension=orange cubes)
- retain indentation on new lines in QR editor
- stop injecting pipe into first command inside a closure
- add `onSelect` callback to autocomplete
- add closure support to `/run {: /echo foo :}`
- add more details to slash command execution exceptions (toasts show where in the script was thrown)
- add option to set scope-specific macros only if a macro with that name does not already exist in the scope (or ancestor scopes)
- add support for simple wildcards in scope-specific macros (`{{arg::*}}`), macros with wildcards are evaluated last
- add `/qr-arg key value` to provide fallback values for missing QR arguments (`/:qrSet.qrLabel key=otherValue`)
- add wildcard `{{arg::*}}` macro on QR execution to replace missing QR arguments without fallback with an empty string
- add util `showFontAwesomePicker()` to show a popup with a list of fa icons and return the selected one
- add font awesome icons to QR buttons (display options: only label, only icon, icon+label)
- add option to transfer (copy / move) a single QR to another QR set
- add option to copy a single QR to clipboard (JSON)
- add option to cut (copy+delete) a single QR to clipboard (JSON)
- add option to save a single QR to JSON file
- add option to paste a single QR into a QR Set from clipboard
- add option add import a single QR into a QR set from a JSON file
- add option to add a new QR / paste QR / import QR in between existing QRs (and before first QR)
- force convert pipe to stringified JSON if it is not one of the allowed types (string, SlashCommandClosure)
- add `/import` to import closures from other QRs (`/import from=MyLib.MyQR foo as bar | /:bar`)
- add referencing QRs by ID (`/qr-update set=myQrSet id=42 newlabel=foo`)
- add `/qr-get` to retrieve a dictionary of a QR's properties
- add optional background / border colors to QR sets
- add block comments `/* stuff *|`, they won't exit early if multiple **complete** block comments (opening and closing delimiters) are nested
- add QR switching dropdown to QR editor (switch to editing another QR)

***

- fix auto complete update on moving caret in textarea
- fix Firefox not showing auto complete without ctrl+space
- fix default value for unnamed args not shown in auto complete command details
- fix `evalBoolean` (numeric comparison if both operands are numbers, otherwise case-insenstive string comparison with non-strings JSON-stringified)
- fix empty strings at the start and end of unsplit but list-parsed unnamed arguments polluting the unnamed argument list
- fix autocomplete not escaping quotes inside quoted values
- fix parser flag REPLACE_GETVAR not working correctly after nested closures
- fix tab / shift-tab in QR editor incorrectly detecting beginning of first line
- fix key listener interference between autocomplete and QR editor
- restore missing methods on popups: `completeAffirmative`, `completeNegative`, `completeCancelled`
- fix `/run` enumProvider not being given current scope
- fix no op when adding QR set to global/chat and first set already included
- fix alphabetical order of QR sets (was case sensitive)

Commit:
don't hide secondary autocomplete values on select or fully typed

Commit:
add scope to enumProvider

Commit:
add autocomplete for multiple unnamed args

Commit:
add enum provider to /add

Commit:
fix whitespace check

Commit:
add base class for name results

Commit:
Merge branch 'staging' into parser-enum-stuff

Commit:
debugger basics rough

Commit:
add default value to unnamed args

Commit:
fix for no arg

Commit:
fixes

Commit:
Merge branch 'staging' into parser-enum-stuff

Commit:
fix firefox and selectionchange issues

Commit:
add matchProvider and valueProvider to autocomplete options

Commit:
coalesce vars in enumProvider

Commit:
use matchProvider and valueProvider in /add arguments

Commit:
debugger stuff

Commit:
track index in getvar replacement

Commit:
add /breakpoint to command list and block it

Commit:
debugger

Commit:
Merge branch 'staging' into parser-enum-stuff

Commit:
debugger

Commit:
debugger

Commit:
indicate pipe, var and getvar command in replaceGetvar

Commit:
indicate immediate closure in toString

Commit:
debugger

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
add missing semicolon

Commit:
add scoped vars and fix hiding (scope->local->global)

Commit:
fix /add multi unnamed

Commit:
don't block enter/tab for items with valueProvider

Commit:
add any var / any number to /add enumProvider

Commit:
debugger

Commit:
trim start of first and end of lat value part in unsplit unnamed arg (analog to pure text arg)

Commit:
only remove if exists

Commit:
debugger

Commit:
add ctrl+click breakpoints

Commit:
fix step while inside subscope

Commit:
highlight /breakpoint

Commit:
fix immediate closure check

Commit:
debugger

Commit:
add count to split unnamed args

Commit:
add split count to /times, /let, and /var

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
add unresolved note

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
fix type checks for evalBoolean and export
- numeric comparison only when both types are numbers
- otherwise use case-insensitive string comparison and JSON-stringify and non-strings

Commit:
export parseBooleanOperands

Commit:
remove SlashCommandClosureExecutor
has long since been replaced with /:

Commit:
add /break to break out of loops

Commit:
add /break support in /times and /while

Commit:
merge staging

Commit:
add command source indicator

Commit:
remove unnecessary escape

Commit:
make /break and /breakpoint show up properly with autocomplete details

Commit:
can't stand this shit any longer

Commit:
show scrollbars with noSyntax

Commit:
restore completeAffirmative, completeNegative, completeCancelled

Commit:
use ctrl+alt+click for breakpoints

Commit:
restore caret after breakpoint click

Commit:
remove empty first string and last string from unsplit list-parsed unnamed argument

Commit:
add missing scopeIndex entries
fixes missing scoped vars in /: auto complete

Commit:
fix run syntax highlight

Commit:
escape quotes inside quoted option value

Commit:
step into closures from elsewhere (draft)

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
fix REPLACE_GETVAR nesting issues

Commit:
make matchProvider and valueProvider optional

Commit:
keep indent on enter

Commit:
fix tab indent line detection

Commit:
fix editor hotkeys and autocomplete interfering

Commit:
add /break value

Commit:
use /break as return statement in /run

Commit:
no pipe inject in first executor of closure

Commit:
update /break help

Commit:
fix key conflicts in QR editor

Commit:
allow options with valueProvider to be selectable

Commit:
add onSelect callback

Commit:
cleanup

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
allow autocomplete on input element (jsdoc)

Commit:
add closure support to /run

Commit:
what?

Commit:
add a little more details to execution exceptions

Commit:
fix enumProvider not getting scope

Commit:
add missing semicolon

Commit:
add setting macro if it not already exists somewhere in scope hierarchy

Commit:
add simple wildcards to scope macros

Commit:
add /qr-arg for {{arg::...}} fallbacks

Commit:
add {{arg::*}} to replace missing args with empty string

Commit:
don't add args._scope etc to macros

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
better stepping into other scripts, with source indicator

Commit:
show source in message label

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
better handling of anonyous source

Commit:
align source right

Commit:
force pipe to stringified JSON if not string or closure

Commit:
add font-awesome picker popup

Commit:
add font-awesome icons to QR buttons

Commit:
more export/import options
- export QR as JSON
- copy QR to clipboard
- cut QR (copy to clipboard and delete)
- transfer QR to other QR Set
- paste QR from clipboard
- import QR from JSON file
- add/paste/import buttons between existing QRs

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
fix no op when adding QR set to global/chat and first set already included

Commit:
add icon support to QR context menu

Commit:
add /import to import closures from other QRs

Commit:
better quoted unnamed arg handling

Commit:
add makeSelectable

Commit:
fix startUnnamedArgs

Commit:
add makeSelectable

Commit:
add referencing QRs by ID

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
add debugger icons

Commit:
fix key listener not updating message

Commit:
jsdoc

Commit:
add syntax highlight for /import

Commit:
fix return for /run with closure

Commit:
fix debug button mask repeat

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
prevent popup close while debugging
breaks shit, but that's a popup problem

Commit:
nowrap in debugger call stack source

Commit:
jsdoc type casing

Commit:
add QR set colors

Commit:
adjust QR set colors for combined buttons

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
clear QR Set color and bg adjustments

Commit:
add minimized debugger

Commit:
remove second step-out button?

Commit:
remove traces of execute on hide

Commit:
fix color without dom

Commit:
add block comments with shortcut and breakpoint shortcut

Commit:
Merge branch 'staging' into parser-followup-2

LenAnderson commented:
Gimme some bugs and issues!

Cohee1207 commented:
> add util showFontAwesomePicker() to show a popup with a list of fa icons and return the selected one

This could be changed to TEXT popup type. CONFIRM type makes no difference here between the options.

Cohee1207 commented:
Weird centering of Show. Nothing align-items center on qr--inputGroup couldn't fix

<img width="561" alt="image" src="https://github.com/user-attachments/assets/c6a0fbab-1828-492e-9b4b-0551f7d0e3d6">

Cohee1207 commented:
faPicker could hide by display none rather than visibility so you wouldn't have to scroll it.

Commit:
fix color picker widget defaulting to black instead of its initial value and firing change event when added to DOM

Commit:
align show label checkbox on baseline

Commit:
use POPUP_TYPE text for fontawesome popup

Commit:
properly hide filtered fontawesime icons in picker

Commit:
Merge branch 'staging' into parser-followup-2

LenAnderson commented:
> faPicker could hide by display none rather than visibility so you wouldn't have to scroll it.

Oh, wow. Had to dig for a bit to understand what you even meant. ST makes `.hidden` = `visibility: hidden;` but Timelines makes `.hidden`= `display: none;`.
Didn't realize that came from an extension and that Timelines just pollutes core classes like that...


POPUP_TYPE, Show alignment, and hidden icons are addressed now.

Commit:
there was a reason I wanted to use confirm, dammit

Commit:
fix QR update/delete by ID

Commit:
more flexibililty for enums
custom mapping from enum value class to enum option class

Commit:
add jsDoc and allow custom icon lists

Commit:
add /pick-icon to show Font Awesome icon picker

Commit:
add icon= and showLabel= to /qr-create and /qr-update

Commit:
add getSetByQr

Commit:
add quick QR switcher to editor

Commit:
fix alphabetical order of QR Sets

Cohee1207 reviewed (approved):
I got no more feedback. Ready when you are.

Commit:
fix toolcool color picker some more

Commit:
add #send_textarea mirror to QR debugger

Commit:
add syntax highlight for pipes and pipe breaks

LenAnderson commented:
@Wolfsblvt Anything left you wanted to test? If not we can merge.

Wolfsblvt commented on public/scripts/extensions/quick-reply/src/QuickReplySet.js:
```diff
@@ -195,10 +240,107 @@ export class QuickReplySet {
+                }
+            }
+            const dlg = new Popup(dom, POPUP_TYPE.CONFIRM, null, { okButton:'Transfer', cancelButton:'Cancel' });
+            const copyBtn = document.createElement('div'); {
```
I think you could just use a `customButton`, which the popup buttons allow. Much less verbose.
If you don't want to have it set its own result value (`2`, for the first custom button), you can give it one manually. If it has a result, it will auto-close the popup. You can also set it to append at the end, and define a click action.

You can either define a button action to set that `isCopy`, or just do it if the popup returns 2.

Much less verbose (:
If you are missing a feature on the buttons. Add it or ask, and I can add it.

LenAnderson replied:
Feel free to change it. But it is neither `appendAtEnd=true` nor `appendAtEnd=false`, the button is added between ok and cancel. So you'd have to do more of them as custom or switch other things around.

Wolfsblvt replied:
It's fine then. Was just a suggestion. Can always be dealt with later once the popup utility expands.

Wolfsblvt committed:
Resize FA icon in QR editor, add label caption

Wolfsblvt committed:
Add 1px margin to qr item adder

Wolfsblvt committed:
Fix keyboard interaction on item adder

Wolfsblvt committed:
Make QR quick buttons aligned

Wolfsblvt committed:
Override/fix width for fa-fw menu_button elements

Wolfsblvt committed:
Fix gap between icon and label, label min width

Wolfsblvt committed:
FA picker popup improvements

- Fixed height (to prevent bouncing and lag/performance issues due to constant resizing. Also looks nicer)
- Added "No icons found" empty text
- Slight rework of button callback

Commit:
Merge branch 'staging' into parser-followup-2

Wolfsblvt commented on public/scripts/slash-commands.js:
```diff
@@ -3481,7 +3518,23 @@ export async function executeSlashCommandsOnChatInput(text, options = {}) {
+                toastr.error(
+                    `${toast}${clickHint}`,
+                    'SlashCommandExecutionError',
+                    { escapeHtml: false, timeOut: 10000, onclick: () => callPopup(toast, 'text') },
```
Should likely not use the legacy popup here, should it?

LenAnderson replied:
just a copy of the parser error handling

Wolfsblvt commented on public/scripts/slash-commands.js:
```diff
@@ -3548,7 +3605,23 @@ async function executeSlashCommandsWithOptions(text, options = {}) {
+                toastr.error(
+                    `${toast}${clickHint}`,
+                    'SlashCommandExecutionError',
+                    { escapeHtml: false, timeOut: 10000, onclick: () => callPopup(toast, 'text') },
```
Same here, this is the legacy popup.

LenAnderson replied:
same

Wolfsblvt reviewed (approved):
None of my feedback, fixes or change requests I consider as blocking for the PR, so I am giving this my approval, but I'd love to see a few things handled still.

### Wants
- [x] Make non-selectable enum values not tab out when pressing tab (and maybe make a visual indication that it's disabled, like grey-isher background
- [x] Change naming of "apply color to border". Like I suggested, "accent"
- [x] Confirm popup delete of QR entries
- [x] Catch and log+toast error of JSON imports from clipboard, and such

### Maybes
- [x] Maybe check import JSON on if its actually one that makes sense there.
- [x] Maybe allow "importing" plaintext or slash commands vai QR clipboard import, just as text
- [x] Hide/disable Debugging buttons in QR editor if you can't use them (just during execution, before any breakpoint has been hit. Otherwise it's confusing)

Mind find more, but those what I noticed today.

Commit:
we have _scope to get the pipe

Commit:
substitute /break argument

Commit:
remove some red

Commit:
add icon to QR switcher "switch sets"

Commit:
add "add QR" to QR switcher

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
fix "Fix keyboard interaction on item adder"

Commit:
faPicker scroll just icons and icon alignment

Commit:
capture Tab key on non-selectable auto complete options

Commit:
change color border -> accent

Commit:
add QR delete confirm

Commit:
verify QR paste JSON and allow non-JSON pastes

Commit:
add debugger button states

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
add autocomplete select key setting (enter/tab)

Wolfsblvt commented:
Don't have time to thoroughly test this before early next week. But from looking at the code seems like all my requests/issues are resolved.

Wolfsblvt commented on public/scripts/extensions/quick-reply/src/QuickReplySet.js:
```diff
@@ -239,6 +239,28 @@ export class QuickReplySet {
+            data = JSON.parse(qrJson ?? '{}');
+            delete data.id;
+        } catch {
+            // not JSON data
```
Not sure if I wouldn't at least still log this to console as an info or error.
Sure, most of the time it might not be JSON, but it could also be malformatted JSON. And to be able to see the reason, having the json parser error with line number might be helpful.
But that's nitpicking again.

Commit:
fix autocomplete setting state on load and duplicated lines

Commit:
fix whitespace variable to "0"

Commit:
fix /let key= not given priority

Commit:
fix unclosed block comment infinite loop

Commit:
don't need regex symbol here

Commit:
improve QR editor performance
- only run hljs with syntax enabled
- only check localStorage once, then rely on the checkbox
- run hljs on a 30fps loop instead of event-based
- use morphdom to update syntax dom instead of innerHTML

Commit:
firefox compat

Commit:
shorter message update debounce

Commit:
use deprecated execCommand to retain undo-history

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
restore closure vars

Succubyss commented:
`/setvar key=x ""` and `/setglobalvar key=x ""` fail in chat but are successful if ran from a QR.

(Also note that toggling the REPLACE_GETVAR setting doesn't cause different outcomes in either case.)

Commit:
there might not be a first or last value in the unnamed arg list

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
rename syntax highlight update loop

Commit:
sync scroll after syntax highlight update

Commit:
don't cancel syntax highlight update if previously disabled

Commit:
don't update command progress CSS property if not changed
performance improvement from luccy

Commit:
await API calls

Commit:
add console warning when delete set is not found in list

Commit:
don't try to parse empty qrJson on add from text

Commit:
add BreakController to /run closure

Commit:
fix some execCommand weirdness

Commit:
hljs refinements

Commit:
adjust debugger highlights for scroll position

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
fix DebugController retaining previous named and unnamed args

Commit:
comments and cleanup

Commit:
check that debugController exists

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
use textarea-owning layer instead of topmost modal layer

Commit:
don't let progress bar go backwards

Commit:
constructor should also allow HTMLInputElement

Commit:
add missing start/end named arg index to /:

Commit:
let /qr-arg overwrite wildcard macros

Commit:
order wildcard macros by index of wildcard for substitute

Commit:
Merge branch 'staging' into parser-followup-2

Commit:
add "for {{input}}" comment to chat input placeholder and title

Commit:
anchor wildcard lookup

LenAnderson commented:
**Overview for users: [V2-2.md](https://github.com/user-attachments/files/16404608/V2-2.md)**

Commit:
fix return value of /delmode

Commit:
fix AbortController missing from imported closures

Commit:
fix syntax loop debugger misalignment

Wolfsblvt committed:
Fix QR context error on no set chosen

Wolfsblvt committed:
QR label alignment/wrapping

Wolfsblvt committed:
add autocomplete color for non-selectable entries

Wolfsblvt committed:
mark popup complete functions as async

Wolfsblvt committed:
Merge branch 'staging' into pr/2415

Commit:
stop chat input and progress bar from collapsing

Commit:
draw attention to debug buttons

Commit:
compile LESS

Commit:
Merge branch 'parser-followup-2' of https://github.com/LenAnderson/SillyTavern into parser-followup-2

Wolfsblvt reviewed (approved):
Finally, my last blessing.

Thanks for adding the tiny changes at last second (:
